WalletWasabi/BranchNBound/BranchAndBound.cs

```diff
@@ -0,0 +1,171 @@
+using NBitcoin;
+using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
+using System.Linq;
+using WalletWasabi.Blockchain.TransactionOutputs;
+using WalletWasabi.Logging;
+
+namespace WalletWasabi.BranchNBound
+{
+	public class BranchAndBound
+	{
+		private readonly Random _random = new();
+
+		public BranchAndBound(List<SmartCoin> availableCoins)
```

Comment:
Use `long` instead of `SmartCoin`.

Reply:
Why? Do you have a use case in mind? 

Then it would be harder to map search results with the original `SmartCoin`s. `long`s are used by the search routine so there is no penalty there. For API consumer, it's also harder to use it correctly then.

Reply:
Because you have to pass what is needed to do the job, no more, no less. BnB is an algorithm to find combinations given numbers that sum at to some number, it is not an algorithm for splitting smartcoins.

The final algorithm that have to find the combination of inputs would require the smartcoins, the network fee rate (to calculate the total effective value (no the `Amount`) and optionally the tolerance and scriptPubKeyType of the output.

Reply:
Thanks. Makes sense.

I was thinking about it a bit more and it occurred to me that we might have a constructor in form:

```cs
public BranchAndBound(Dictionary<T, long> availableCoins)
```

not to lose the mapping between an object and its value but then given that results of BnB over longs can be quite easily mapped back to whatever source of data, it does not seem that useful.